README
This repository contains a hardware prefetcher based on a basic cache simulator written in C/C++. And simulated memory hierarchy is described as below:
- L1 Data Cache
  - 32KB Size
  - 2-way set associative
  - 16 byte line size
  - write-through
  - no-write-allocate
  - 1 cycle hit time
  - 10 entry write-buffer
- L2 Cache (data only)
  - 256KB Size
  - 8-way set associative
  - 32 bytes line size
  - write-back
  - write-allocate
  - 20 cycle access time (fully pipelined)
- Main Memory
  - 50 cycle access time (not pipelined)
  - accesses to main memory are held in a FIFO queue with 10 entries
  - L2 cache will no longer accept requests if the FIFO is full


src_naive/changelog
Version 1.6 (Feb 15, 2009):
- Fixed bug where the simulator will hang in some benchmarks.

Version 1.5 (Dec 3, 2007):
- Fixed bug where store request coming from memory queue would prematurely
end a CPU load request (sorry, but this fix often decreases performance).
- Fixed a bug where stores coming from memory queue were updating the
D-cache erroneously.
- setStatus() in CPU class doesn't return anything so it is now declared
to return void.

Version 1.4 (Nov 28, 2007):
- Fixed bug in main.C where stale prefetcher requests rather than the
desired CPU request were getting added to L2 queue after entering 
STALLED_L2 state.

Version 1.3 (Nov 22, 2007):
- Added check in main.C to make sure all requests from the prefetcher are
labeled as loads.

Version 1.2 (Nov 20, 2007):
- Fixed bug in readNextRequest() of CPU that marked all accesses as loads.

Version 1.1 (Nov 18, 2007):
- Fixed bug in findDup() function of memQueue.
- Added printQueue() helper function.

Version 1.0:
- Initial Version


src_hybrid/prefetcher.C
/**
 * Name:  Yunqi Zhang
 * PID:   A53030068
 * Email: yqzhang@ucsd.edu
 */

#include "prefetcher.h"
#include "mem-sim.h"

Prefetcher::Prefetcher() {
  initLocalRequest();
  initHistoryState();
}

bool Prefetcher::hasRequest(u_int32_t cycle) {
	return (!ifEmptyLocalRequest());
}

Request Prefetcher::getRequest(u_int32_t cycle) {
	Request req;
  req.addr = getFrontLocalRequest();
	return req;
}

void Prefetcher::completeRequest(u_int32_t cycle) {
  deLocalRequest();
}

void Prefetcher::cpuRequest(Request req) {
  u_int16_t index = queryHistoryState(req.pc);

  // Not in the history state table
  if (index == NULL_STATE) {
    insertHistoryState(req.pc, req.addr, 1, DEFAULT_OFFSET, DEFAULT_AHEAD, STATE_INIT);
    sequentialPrefetch(req.addr, DEFAULT_OFFSET, DEFAULT_AHEAD, req.HitL1);
  }
  // Already in the history state table
  else {
    int16_t offset = (int64_t)req.addr - (int64_t)historyState[index].addr;
      
    switch (historyState[index].state) {
    case STATE_INIT:
      if (offset == historyState[index].offset) {
        historyState[index].state = STATE_STEADY;
        historyState[index].addr  = req.addr;
        historyState[index].count ++;
        historyState[index].ahead --;
        stridePrefetch(req.addr, offset, historyState[index].count, historyState[index].ahead, req.HitL1);
      }
      else {
        historyState[index].state = STATE_TRANS;
        historyState[index].addr  = req.addr;
        historyState[index].count = 1;
        historyState[index].offset= offset;
        historyState[index].ahead = DEFAULT_AHEAD;
        sequentialPrefetch(req.addr, NOT_HIT_OFFSET, NOT_HIT_AHEAD, req.HitL1);
      }
      break;
    case STATE_TRANS:
      if (offset == historyState[index].offset) {
        historyState[index].state = STATE_STEADY;
        historyState[index].addr  = req.addr;
        historyState[index].count = 2;
        historyState[index].ahead --;
        stridePrefetch(req.addr, offset, historyState[index].count, historyState[index].ahead, req.HitL1);
      }
      else {
        historyState[index].state = STATE_NO_PRE;
        historyState[index].addr  = req.addr;
        historyState[index].count = 1;
        historyState[index].offset= offset;
        historyState[index].ahead = DEFAULT_AHEAD;
        sequentialPrefetch(req.addr, NOT_HIT_OFFSET, NOT_HIT_AHEAD, req.HitL1);
      }
      break;
    case STATE_STEADY:
      if (offset == historyState[index].offset) {
        historyState[index].state = STATE_STEADY;
        historyState[index].addr  = req.addr;
        historyState[index].count += 4;
        historyState[index].ahead --;
        stridePrefetch(req.addr, offset, historyState[index].count, historyState[index].ahead, req.HitL1);
      }
      else {
        historyState[index].state = STATE_TRANS;
        historyState[index].addr  = req.addr;
        historyState[index].offset= offset;
        historyState[index].count = 1;
        historyState[index].ahead = DEFAULT_AHEAD;
        sequentialPrefetch(req.addr, NOT_HIT_OFFSET, NOT_HIT_AHEAD, req.HitL1);
      }
      break;
    case STATE_NO_PRE:
      if (offset == historyState[index].offset) {
        historyState[index].state = STATE_TRANS;
        historyState[index].addr  = req.addr;
        historyState[index].count = 2;
        historyState[index].ahead --;
        sequentialPrefetch(req.addr, offset, historyState[index].ahead, req.HitL1);
      }
      else {
        historyState[index].addr  = req.addr;
        historyState[index].count = 1;
        historyState[index].offset= offset;
        historyState[index].ahead = DEFAULT_AHEAD;
        sequentialPrefetch(req.addr, NOT_HIT_OFFSET, NOT_HIT_AHEAD, req.HitL1);
      }
      break;
    default:
      printf("An error occurred in the finite state machine T_T\n");
    }

    //Place to the most recent used head
    if (historyState[index].next == NULL_STATE) {
      u_int16_t prev = getSecondLRUState();

      historyState[prev].next = NULL_STATE;
      historyState[index].next = stateHead;
      stateHead = index;
    }
    else {
      u_int16_t next = historyState[index].next;

      State temp;
      temp.pc     = historyState[index].pc;
      temp.addr   = historyState[index].addr;
      temp.count  = historyState[index].count;
      temp.offset = historyState[index].offset;
      temp.ahead  = historyState[index].ahead;
      temp.state  = historyState[index].state;

      historyState[index].pc      = historyState[next].pc;
      historyState[index].addr    = historyState[next].addr;
      historyState[index].count   = historyState[next].count;
      historyState[index].offset  = historyState[next].offset;
      historyState[index].ahead   = historyState[next].ahead;
      historyState[index].state   = historyState[next].state;
      historyState[index].next    = historyState[next].next;

      historyState[next].pc     = temp.pc;
      historyState[next].addr   = temp.addr;
      historyState[next].count  = temp.count;
      historyState[next].offset = temp.offset;
      historyState[next].ahead  = temp.ahead;
      historyState[next].state  = temp.state;
      historyState[next].next   = stateHead;
      stateHead = next;
    }
  }
}

// History state table operations
void Prefetcher::initHistoryState() {
  stateCount = 0;
  stateHead = NULL_STATE;
}

bool Prefetcher::ifEmptyHistoryState() {
  return (stateCount == 0);
}

bool Prefetcher::ifFullHistoryState() {
  return (stateCount == MAX_STATE_COUNT);
}

void Prefetcher::insertHistoryState(u_int32_t pc, u_int32_t addr, u_int16_t count, int16_t offset, u_int16_t ahead, u_int8_t state) {
  if (!ifFullHistoryState()) {
    // Insert state when table is not full
    historyState[stateCount].pc     = pc;
    historyState[stateCount].addr   = addr;
    historyState[stateCount].count  = count;
    historyState[stateCount].offset = offset;
    historyState[stateCount].ahead  = ahead;
    historyState[stateCount].next   = stateHead;
    historyState[stateCount].state  = state;
    
    stateHead = stateCount;
    stateCount++;
  }
  else {
    // LRU eviction in history state table
    u_int32_t index = getSecondLRUState();
    u_int32_t replace = historyState[index].next;
    historyState[index].next = NULL_STATE;

    historyState[replace].pc      = pc;
    historyState[replace].addr    = addr;
    historyState[replace].count   = count;
    historyState[replace].offset  = offset;
    historyState[replace].ahead   = ahead;
    historyState[replace].next    = stateHead;
    historyState[replace].state   = state;

    stateHead = replace;
  }
}

u_int16_t Prefetcher::queryHistoryState(u_int32_t pc) {
  u_int16_t i;
  for (i = stateHead; i != NULL_STATE; i = historyState[i].next ) {
    if (historyState[i].pc == pc) {
      return i;
    }
  }
  return NULL_STATE;
}

u_int16_t Prefetcher::getSecondLRUState() {
  u_int16_t i;
  for (i = 0; i < MAX_STATE_COUNT; i++) {
    if (historyState[i].next == NULL_STATE) {
      continue;
    }
    else if (historyState[historyState[i].next].next == NULL_STATE) {
      return i;
    }
  }
  printf("Error -- from Prefetcher::getSecondLRUState()");
}

// Local request queue operations
void Prefetcher::initLocalRequest() {
  frontRequest = rearRequest = 0;
}

bool Prefetcher::ifEmptyLocalRequest() {
  return (frontRequest == rearRequest);
}

bool Prefetcher::ifFullLocalRequest() {
  return (frontRequest == (rearRequest + 1) % MAX_REQUEST_COUNT);
}

bool Prefetcher::enLocalRequest(u_int32_t addr, bool HitL1) {
  if (ifAlreadyInQueue(addr, HitL1)) {
    return false;
  }
  if (ifFullLocalRequest()) {
    return false;
  }
  else {
    localRequest[rearRequest] = addr;
    rearRequest = (rearRequest + 1) % MAX_REQUEST_COUNT;
    return true;
  }
}

u_int32_t Prefetcher::deLocalRequest() {
  if (ifEmptyLocalRequest()) {
    return 0;
  }
  else {
    u_int32_t addr = localRequest[frontRequest];
    frontRequest = (frontRequest + 1) % MAX_REQUEST_COUNT;
    return addr;
  }
}

u_int32_t Prefetcher::getFrontLocalRequest() {
  return localRequest[frontRequest];
}

bool Prefetcher::ifAlreadyInQueue(u_int32_t addr, bool HitL1) {
  u_int32_t i;
  for (i = frontRequest; i % MAX_REQUEST_COUNT < rearRequest; i++) {
    if (HitL1) {
      if (localRequest[i % MAX_REQUEST_COUNT] / L1_CACHE_BLOCK == addr / L1_CACHE_BLOCK) {
        return true;
      }
    }
    else {
      if (localRequest[i % MAX_REQUEST_COUNT] / L2_CACHE_BLOCK == addr / L2_CACHE_BLOCK) {
        return true;
      }
    }
  }
  return false;
}

void Prefetcher::sequentialPrefetch(u_int32_t addr, int16_t offset, u_int16_t ahead, bool HitL1) {
  int16_t i;
  if (offset != 0) {
    for (i = -1; i <= ahead; i++) {
      u_int32_t prefetch;
      prefetch = (int64_t)addr + (int64_t)i * (int64_t)offset;
      if (HitL1) {
        if (prefetch / L1_CACHE_BLOCK == addr / L1_CACHE_BLOCK) continue;
      }
      else {
        if (prefetch / L2_CACHE_BLOCK == addr / L2_CACHE_BLOCK) continue;
      }
      enLocalRequest(prefetch, false);
    }  
  }
}

void Prefetcher::stridePrefetch(u_int32_t addr, int16_t offset, u_int16_t count, u_int16_t &ahead, bool HitL1) {
  if (offset != 0) {
    while (ahead < MIN(count, MAX_PREFETCH_DEGREE)) {
      ahead++;
      enLocalRequest((int64_t)addr + (int64_t)ahead * (int64_t)offset, HitL1);
    }
  }
}


src_hybrid/prefetcher.h
/**
 * Name:  Yunqi Zhang
 * PID:   A53030068
 * Email: yqzhang@ucsd.edu
 */

#ifndef PREFETCHER_H
#define PREFETCHER_H

#define MAX_STATE_COUNT 64
#define MAX_REQUEST_COUNT 64
#define NULL_STATE 0xFF
#define L1_CACHE_BLOCK 16
#define L2_CACHE_BLOCK 32

#define DEFAULT_OFFSET 32
#define DEFAULT_AHEAD  2

#define NOT_HIT_OFFSET 16
#define NOT_HIT_AHEAD  2

#define MAX_PREFETCH_DEGREE 16

#define STATE_INIT   0
#define STATE_TRANS  1
#define STATE_STEADY 2
#define STATE_NO_PRE 3

#define MAX(x, y) (((x) > (y)) ? (x) : (y))
#define MIN(x, y) (((x) < (y)) ? (x) : (y))
#define ABS(x)    (((x) < 0) ? -(x) : (x))

#include <sys/types.h>
#include <stdio.h>

struct Request;

struct State {
  u_int32_t pc;     // PC of the state      - 32 bit
  u_int32_t addr;   // Last access address  - 32 bit
  u_int16_t count;  // Access counter       - 16 bit
  int16_t offset;   // Access offset        - 16 bit
  u_int16_t ahead;  // Prefetching ahead    - 16 bit
  u_int8_t next;   // Next state for LRU   - 16 bit
  u_int8_t state;   // "state" of state     - 8  bit
};

class Prefetcher {
  private:
  // History state table - Size: sizeof(State) * MAX_STATE_COUNT
  u_int32_t stateCount;
  u_int32_t stateHead;
  State historyState[MAX_STATE_COUNT];

  // Local request queue
  u_int32_t frontRequest;
  u_int32_t rearRequest;
  u_int32_t localRequest[MAX_REQUEST_COUNT];

  // History state table operations
  void initHistoryState();
  bool ifEmptyHistoryState();
  bool ifFullHistoryState();
  void insertHistoryState(u_int32_t, u_int32_t, u_int16_t, int16_t, u_int16_t, u_int8_t);
  u_int16_t queryHistoryState(u_int32_t);
  u_int16_t getSecondLRUState();

  // Local request queue operations
  void initLocalRequest();
  bool ifEmptyLocalRequest();
  bool ifFullLocalRequest();
  bool enLocalRequest(u_int32_t, bool);
  u_int32_t deLocalRequest();
  u_int32_t getFrontLocalRequest();
  bool ifAlreadyInQueue(u_int32_t, bool);

  // Prefetch algorithm
  void sequentialPrefetch(u_int32_t, int16_t, u_int16_t, bool);
  void stridePrefetch(u_int32_t, int16_t, u_int16_t, u_int16_t&, bool);

  public:
  // Construction function
  Prefetcher();

	// should return true if a request is ready for this cycle
	bool hasRequest(u_int32_t cycle);

	// request a desired address be brought in
	Request getRequest(u_int32_t cycle);

	// this function is called whenever the last prefetcher request was successfully sent to the L2
	void completeRequest(u_int32_t cycle);

	/*
	 * This function is called whenever the CPU references memory.
	 * Note that only the addr, pc, load, issuedAt, and HitL1 should be considered valid data
	 */
	void cpuRequest(Request req); 
};

#endif


src_naive/main.C
/*
 *
 * File: main.C
 * Author: Saturnino Garcia (sat@cs)
 * Description: Cache simulator for 2 level cache
 *
 */

#include "cache.h"
#include "CPU.h"
#include "mem-sim.h"
#include "memQueue.h"
#include "prefetcher.h"

#include <stdlib.h>
#include <string.h>

int main(int argc, char* argv[]) {
	if(argc != 2) {
		printf("Usage: %s [trace file]\n",argv[0]);
		return 1;
	}

	int hitTimeL1 = 1; 
	int accessTimeL2 = 20; 
	int accessTimeMem = 50; 

	int lineSizeL1 = 16;
	int assocL1 = 2; 
	int totalSizeL1 = 32; 

	int lineSizeL2 = 32; 
	int assocL2 = 8; 
	int totalSizeL2 = 256; 

	int numSetsL1, numSetsL2;


	u_int32_t addr, cycles;
	u_int32_t runtime = 0; // number of cycles in runtime
	u_int32_t nonmemrt = 0; // number of cycles in runtime spent with non-mem instructions

	FILE *fp; // to write out the final stats

	// calc number of sets (if assoc == 0 then it's fully assoc so there is only 1 set)
	if(assocL1 != 0) numSetsL1 = totalSizeL1 * 1024 / (assocL1 * lineSizeL1);
	else {
		numSetsL1 = 1;
		assocL1 = totalSizeL1 * 1024 / lineSizeL1;
	}

	if(assocL2 != 0) numSetsL2 = totalSizeL2 * 1024 / (assocL2 * lineSizeL2);
	else {
		numSetsL2 = 1;
		assocL2 = totalSizeL2 * 1024 / lineSizeL2;
	}

	// D-cache is write through with no-write-alloc, LRU replacement
	Cache DCache(numSetsL1,assocL1,lineSizeL1,false,false,true);

	// L2 cache is writeback with write-alloc, LRU replacement
	Cache L2Cache(numSetsL2,assocL2,lineSizeL2,false,true,false);

	CPU cpu(argv[1]);

	Prefetcher pf;

	memQueue writeBuffer(10,&DCache,accessTimeL2,true,true,'a');
	memQueue queueL2(20,&DCache,accessTimeL2,true,false,'b');
	memQueue queueMem(10,&L2Cache,accessTimeMem,false,false,'c');


	// statistical stuff
	u_int32_t nRequestsL2 = 0; // number of requests sent out to L2 (both CPU and prefetcher requests)
	u_int32_t memCycles = 0; // number of cycles that main memory is being accessed
	u_int32_t memQsize = 0; // used for calculating average queue length

	u_int32_t curr_cycle = 1;
	Request req;
	bool isHit;

	while(!cpu.isDone()) {

		isHit = false;

		cpuState cpu_status = cpu.getStatus(curr_cycle);

//		printf("%u: %u\n",curr_cycle,cpu_status);

		if(cpu_status == READY) { // request is ready
				
			req = cpu.issueRequest(curr_cycle);

			// check for L1 hit
			isHit = DCache.check(req.addr,req.load);
			cpu.hitL1(isHit);
			req.HitL1 = isHit;

			// notify the prefetcher of what just happened with this memory op
			pf.cpuRequest(req, &DCache, &L2Cache);

			if(isHit) {
				DCache.access(req.addr,req.load);
				cpu.completeRequest(curr_cycle);
			}
			else if(req.load) {
				nRequestsL2++;
				if(queueL2.add(req,curr_cycle)) cpu.setStatus(WAITING); // CPU is now "waiting" for response from L2/mem
				else cpu.setStatus(STALLED_L2); // no room in l2 queue so we are "stalled" on this request
			}
			else { 
				nRequestsL2++;

				if (writeBuffer.add(req,curr_cycle)) cpu.completeRequest(curr_cycle); 
				else { // need to stall for an entry in the write buffer to open up
					cpu.setStatus(STALLED_WB);
				}
			}
		}
		// PF can do some work if we are just waiting or idle OR if we had a hit in the D-cache so the D-to-L2 bus isn't needed
		else if(cpu_status == WAITING || cpu_status == IDLE || cpu_status == STALLED_WB || isHit) { // either waiting for lower mem levels or idle so PF can do something
			if(pf.hasRequest(curr_cycle)) { 
				nRequestsL2++;

				req = pf.getRequest(curr_cycle);
				req.fromCPU = false;
				req.load = true;
				if(queueL2.add(req,curr_cycle)) pf.completeRequest(curr_cycle); // if added to queue then the request is "complete"
			}

			if(cpu_status == STALLED_WB) { // attempt to put it in the write buffer
				req = cpu.getRequest(); // get the request we want

				if (writeBuffer.add(req,curr_cycle)) cpu.completeRequest(curr_cycle); // if added, we can move on
			}
		}
		else if(cpu_status == STALLED_L2) { // stalled b/c of L2 queue so let us just try this right away
			req = cpu.getRequest();
			if(queueL2.add(req,curr_cycle)) cpu.setStatus(WAITING); // l2 queue is free now so we can go into waiting state
		}

		// service the L2 queue
		if(queueL2.frontReady(curr_cycle)) { // check to see if the front element in the queue is ready
			//printf("servicing the l2 queue on cycle %u\n",curr_cycle);
			req = queueL2.getFront();

			isHit = L2Cache.check(req.addr,req.load);
			cpu.loadHitL2(isHit);

			if(isHit) {
				DCache.access(req.addr,req.load); // update D cache
				if(req.fromCPU) cpu.completeRequest(curr_cycle); // this request was from the CPU so update state to show we are done
				queueL2.remove(); // remove this request from the queue
			}
			else {
				if(queueMem.add(req,curr_cycle)) queueL2.remove(); // succesfully added to memory queue so we can remove it from L2 queue
			}
		}

		// service the memory queue
		if(queueMem.frontReady(curr_cycle)) {
			//printf("servicing the mem queue on cycle %u\n",curr_cycle);
			req = queueMem.getFront();
			queueMem.remove();

			// update both L2 and D cache
			L2Cache.access(req.addr,req.load);
			if(req.load) DCache.access(req.addr,req.load); // only update if this is a load

			if(req.fromCPU && req.load) cpu.completeRequest(curr_cycle);
		}

		// check to see if we are utilizing memory BW during this cycle
		if(queueMem.getSize() > 0) memCycles++;

		// used to find the average size of the memory queue
		memQsize += queueMem.getSize();

		// service the write buffer
		if(writeBuffer.frontReady(curr_cycle)) {
			req = writeBuffer.getFront();

			isHit = L2Cache.check(req.addr,req.load);
			cpu.storeHitL2(isHit);

			if(isHit) { // store hit in L2 so just save it and we are done
				L2Cache.access(req.addr,req.load);
				writeBuffer.remove();
			}
			else { // L2 is write-allocate so we need to load data from memory first
				if(queueMem.add(req,curr_cycle)) writeBuffer.remove(); // we can keep adding to the queue because we check for duplicates as part of add()
			}
		}

		curr_cycle++; // next cycle
	}

	curr_cycle--; // just for stats sake

	double avgMemQ = (double)memQsize / (double)curr_cycle;
	double L2BW = (double)nRequestsL2 / (double)curr_cycle;
	double memBW = (double)memCycles / (double)curr_cycle;

	/*
	fprintf("total run time: %u\n",curr_cycle);
	fprintf("D-cache total hit rate: %f\n",cpu.getHitRateL1());
	fprintf("L2 cache total hit rate: %f\n",cpu.getHitRateL2());
	fprintf("AMAT: %f\n",cpu.getAMAT());
	fprintf("Average Memory Queue Size: %f\n",avgMemQ);
	fprintf("L2 BW Utilization: %f\n",L2BW);
	fprintf("Memory BW Utilization: %f\n",memBW);
	*/


	// create output file name based on trace file name
	char* outfile = (char *)malloc(sizeof(char)*(strlen(argv[1])+5));
	strcpy(outfile,argv[1]);
	strcat(outfile,".out");

	fp = fopen(outfile,"w"); // open outfile for writing

	free(outfile);

	fprintf(fp,"%u\n",curr_cycle);
	fprintf(fp,"%.4f\n",cpu.getHitRateL1());
	fprintf(fp,"%.4f\n",cpu.getHitRateL2());
	fprintf(fp,"%.4f\n",cpu.getAMAT());
	fprintf(fp,"%.4f\n",avgMemQ);
	fprintf(fp,"%.4f\n",L2BW);
	fprintf(fp,"%.4f\n",memBW);

	fclose(fp);

	return 0;
}

